controlplane/internal/controller/imageregistry/image_registry.go

```diff
@@ -0,0 +1,199 @@
+// a ConfigMap containing the additional trusted certificates for the mirror registry, and an image.config.openshift.io
+// CR that references the ConfigMap with the additional trusted certificate.
+// Successful creation of the ConfigMap will return the name of the created ConfigMap to be added as a manifest reference
+func CreateConfig(ctx context.Context, client client.Client, registryRef *corev1.LocalObjectReference, namespace string) (string, error) {
+
+	imageRegistryConfigMap := &corev1.ConfigMap{}
+	if err := client.Get(ctx, types.NamespacedName{Name: registryRef.Name, Namespace: namespace}, imageRegistryConfigMap); err != nil {
+		return "", err
+	}
+
+	additionalRegistryConfigMap, err := createImageRegistryConfig(imageRegistryConfigMap, namespace)
+	if err != nil {
+		return "", err
+	}
+
+	if err := client.Create(ctx, additionalRegistryConfigMap); err != nil {
+		return "", err
+	}
+	return additionalRegistryConfigMap.Name, nil
+}
+
+func createImageRegistryConfig(imageRegistry *corev1.ConfigMap, namespace string) (*corev1.ConfigMap, error) {
+	imageConfig := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      imageConfigMapName,
+			Namespace: namespace,
+		},
+		Data: make(map[string]string),
+	}
+	registryConf, ok := imageRegistry.Data[registryConfKey]
+	if ok {
+		imageDigestMirrorSet, err := getImageRegistries(registryConf)
+		if err != nil {
+			return nil, err
+		}
+		imageConfig.Data[imageDigestMirrorSetKey] = imageDigestMirrorSet
+	}
+
+	registryCert, ok := imageRegistry.Data[registryCertKey]
+	if ok {
+		addImageRegistryCertificate(imageConfig, registryCert)
+	}
+	return imageConfig, nil
+}
+
+// getImageRegistries reads a toml tree string with the structure:
+//
+// [[registry]]
+//
+//	location = "source-registry"
+//
+//	 [[registry.mirror]]
+//	 location = "mirror-registry"
+//
+// and converts it to an ImageDigestMirrorSet CR
+// and returns it as a marshalled JSON string
+func getImageRegistries(registry string) (string, error) {
+	// Parse toml and add mirror registry to image digest mirror set
+	tomlTree, err := toml.Load(registry)
+	if err != nil {
+		return "", err
+	}
+
+	idmsMirrors := make([]configv1.ImageDigestMirrors, 0)
+	registriesTree, ok := tomlTree.Get("registry").([]*toml.Tree)
+	if !ok {
+		return "", fmt.Errorf("failed to find registry in toml tree")
+	}
+
+	// Add each registry and its mirrors as a image digest mirror object
+	for _, registryTree := range registriesTree {
+		var imageDigestMirror configv1.ImageDigestMirrors
+		source, ok := registryTree.Get("location").(string)
+		if !ok {
+			continue
+		}
+		imageDigestMirror.Source = source
+
+		var mirrors []configv1.ImageMirror
+		mirrorTrees, ok := registryTree.Get("mirror").([]*toml.Tree)
+		if !ok {
+			continue
+		}
+
+		for _, mirrorTree := range mirrorTrees {
+			mirror, ok := mirrorTree.Get("location").(string)
+			if ok {
+				mirrors = append(mirrors, configv1.ImageMirror(mirror))
+			}
```

Comment:
I think inlining it would be tidier
```if mirror, ok := mirrorTree.Get("location").(string); ok {
    mirrors = append(mirrors, configv1.ImageMirror(mirror))
}
```

Reply:
Good catch, thanks!